{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Correlation Analysis\n",
    "\n",
    "In bioinformatics, correlation can be used to identify coregulated gene expression, or identify an association between gene expression and effect of a drug. We will explore both in this example notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Import Libraries we need for our analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import string\n",
    "import random\n",
    "import pprint\n",
    "import json\n",
    "import time\n",
    "import os\n",
    "pp = pprint.PrettyPrinter(width=79, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define your credentials\n",
    "\n",
    "Credentials can be found on [your Lumin Workspace page](https://database.championsoncology.com/lumin/workstation/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define user credentials\n",
    "username = '5036-99-ca8c10'\n",
    "client_id = 99\n",
    "user_id = 5036\n",
    "token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwZXJtaXNzaW9uIjoid29ya3N0YXRpb24iLCJpYXQiOjE2NDAxMjUyNzUsImV4cCI6MTY0MDEzOTY3NX0.CYUw1P9vIkHPLtm95fYbzY2nKedh3ecNpTljH-lHQxo'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is a function we use later to retrieve data. We use this to abstract away the work of a data request. You shouldn't need to change anything here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def request_data(table=\"expression\", cancer_type = [\"all\"], genes = [], samples = [], drugs = None, source = \"PDX\", data_set_name = None):\n",
    "    \n",
    "    if data_set_name is None:\n",
    "        # generate a folder name if none was specified\n",
    "        res = ''.join(random.choices(string.ascii_uppercase +\n",
    "                             string.digits, k = 6))\n",
    "        data_set_name = 'Data-'+res\n",
    "    else:\n",
    "        # Remove any special characters\n",
    "        data_set_name = ''.join(e for e in data_set_name if e.isalnum())\n",
    "\n",
    "    # Define the data request dictionary\n",
    "    d = {\n",
    "        \"request_data_type\": table,\n",
    "        \"request_cancer_type\": cancer_type,\n",
    "        \"request_genes\": genes,\n",
    "        \"request_models\": samples,\n",
    "        \"request_agents\": None,\n",
    "        \"request_dataset\": source,\n",
    "        \"request_workspace_id\": username,\n",
    "        \"request_client\": client_id,\n",
    "        \"request_user\": user_id,\n",
    "        \"request_mode\": True,\n",
    "        \"request_display_error\": False,\n",
    "        \"preview\": True\n",
    "    }\n",
    "\n",
    "    # create the request\n",
    "    headers={\"authorization\":\"Bearer \"+token}\n",
    "    response = requests.post('https://stag.lumin-fast-api.championsoncology.com/workstation/',json=d, headers=headers)\n",
    "    task_id = None\n",
    "    task_id = response.json()['task_id']\n",
    "\n",
    "    if task_id is None:\n",
    "        print(\"Error: No task_id returned. Please check the request details\")\n",
    "        return None\n",
    "    else:\n",
    "        print('')\n",
    "        print(\"Data Request ID: {}\".format(task_id))\n",
    "\n",
    "        # check the status of the request until it's complete\n",
    "        while True:\n",
    "            response = requests.get('https://stag.lumin-fast-api.championsoncology.com/tasks/poll/'+task_id+'?old_progress=0',json=d, headers=headers)\n",
    "            state = response.json()['state']\n",
    "            if state == 'SUCCESS':\n",
    "                result = response.json()['result']\n",
    "                rd = json.loads(result)\n",
    "                #pp.pprint(rd)\n",
    "                folder = rd['name']\n",
    "                break\n",
    "            elif state == 'PROGRESS':\n",
    "                print(\"Still loading...\")\n",
    "            elif state == 'error':\n",
    "                print(\"Error: {}\".format(response.json()['error']))\n",
    "                return None, None\n",
    "            else:\n",
    "                print(\"Status: {}\".format(state))\n",
    "                time.sleep(20)\n",
    "            #pp.pprint(response.json())\n",
    "\n",
    "        # Change the folder name to the data set name\n",
    "        directory = os.path.expanduser(\"~\")\n",
    "        #print(directory)\n",
    "        os.rename(directory+'/'+folder, directory+'/'+data_set_name)\n",
    "\n",
    "        # combine the files and save as a CSV\n",
    "        folder = directory+\"/\"+data_set_name\n",
    "        data = pd.DataFrame()\n",
    "        i = 0\n",
    "        for file in os.listdir(folder):\n",
    "            i = i+1\n",
    "            print(f'reading file {i} {file} ...')\n",
    "            df = pd.read_json(folder+'/'+file)\n",
    "            data = pd.concat([data, df])\n",
    "    \n",
    "        #print(data)\n",
    "        data.to_csv(folder+'/'+data_set_name+'.csv', encoding='utf-8')\n",
    "\n",
    "        # return the file name and location to the user\n",
    "        return data,folder+'/'+data_set_name+'.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Request Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Request TGI data - run once\n",
    "#df, location = request_data(table=\"TGI\", drugs=[\"Olaparib\"])\n",
    "#print(\"Data saved to {} and available as df\".format(location))\n",
    "\n",
    "# Save just the columns we need\n",
    "#data_df = df[['model_name','agent','TGI','tumor_type']]\n",
    "#print(data_df)\n",
    "\n",
    "#data_df.to_csv(location, encoding='utf-8')\n",
    "\n",
    "\n",
    "# Load from CSV next time\n",
    "location = '../Data/TGI.csv'\n",
    "load_data_df = pd.read_csv(location)\n",
    "print(load_data_df.head(5))\n",
    "\n",
    "#df, location = request_data(table=\"expression\", genes = ['BRCA1', 'BRCA2'], data_set_name=\"BRCA!_BRCA2\")\n",
    "#print(\"Data saved to {} and available as df\".format(location))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.1 Define the data we want"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get distinct models or drugs\n",
    "unique_samples = load_data_df.agent.unique().tolist()\n",
    "print(unique_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get selected rows\n",
    "selected_rows = load_data_df.loc[load_data_df['agent'] == 'Olaparib']\n",
    "print(selected_rows)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.2 Make our request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the models we are interested in as a list\n",
    "unique_samples = selected_rows.model_name.unique().tolist()\n",
    "#print(unique_samples)\n",
    "#df, location = request_data(table=\"expression\", genes = [], samples = unique_samples, data_set_name=\"Olaparib_expression\")\n",
    "#print(\"Data saved to {} and available as df\".format(location))\n",
    "\n",
    "location = '../Data/Olaparib_expression.csv'\n",
    "load_data_df = pd.read_csv(location)\n",
    "print(load_data_df.head(5))\n",
    "\n",
    "exp_df = load_data_df[['model_name','log.tpm','tumor_type','gene']]\n",
    "print(exp_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a matrix\n",
    "pivot_df = exp_df.reset_index().pivot_table(index=\"model_name\", columns=\"gene\", values=\"log.tpm\", aggfunc='mean')\n",
    "print(pivot_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Keep only a random 100 genes so the example runs fast and we can plot the results\n",
    "small_pivot_df = pivot_df.sample(n=40,axis='columns')\n",
    "\n",
    "corr = small_pivot_df.corr()\n",
    "plt.figure(figsize = (16,16))\n",
    "ax = sns.heatmap(\n",
    "    corr, \n",
    "    vmin=-1, vmax=1, center=0,\n",
    "    cmap=sns.diverging_palette(20, 220, n=200),\n",
    "    square=True\n",
    ")\n",
    "ax.set_xticklabels(\n",
    "    ax.get_xticklabels(),\n",
    "    rotation=45,\n",
    "    horizontalalignment='right'\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add Olaparib Response data\n",
    "\n",
    "# Add index to values_df\n",
    "new_small_pivot_df = small_pivot_df.reset_index()\n",
    "#print(new_small_pivot_df.shape)\n",
    "\n",
    "TGI_data = selected_rows.groupby(['model_name']).mean()\n",
    "#print(TGI_data.shape)\n",
    "\n",
    "tgi_values = TGI_data['TGI'].tolist()\n",
    "corr_data = new_small_pivot_df\n",
    "\n",
    "corr_data['TGI'] = tgi_values[:-1]\n",
    "#print(corr_data)\n",
    "\n",
    "#TGI_corr_results = new_small_pivot_df.corrwith(TGI_data['TGI'], axis='columns', drop=True, method='pearson')\n",
    "#print(TGI_corr_results)\n",
    "results = corr_data[corr_data.columns[:]].corr()['TGI'][:-1]\n",
    "print(results)\n",
    "\n",
    "#plt.figure(figsize = (16,16))\n",
    "##ax = sns.heatmap(\n",
    "#    results, \n",
    "#    vmin=-1, vmax=1, center=0,\n",
    "#    cmap=sns.diverging_palette(20, 220, n=200),\n",
    "#    square=True\n",
    "#)\n",
    "#ax.set_xticklabels(\n",
    "#    ax.get_xticklabels(),\n",
    "#    rotation=45,\n",
    "#    horizontalalignment='right'\n",
    "#);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.7.3 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
